### PR TITLE
Implement new Pester Tests for Visual Studio

### PR DESCRIPTION
**Pull request by dibir-magomedsaygitov**
# Description
Implement new Pester Tests for Visual Studio.
Also we have made following changes in the list of VS components:

**VS2017:**
Added | Not available
-- | --  
Microsoft.VisualStudio.Component.Sharepoint.Tools | Microsoft.VisualStudio.Component.TeamOffice.BuildTools
Microsoft.VisualStudio.Component.TeamOffice | Microsoft.VisualStudio.Component.Sharepoint.BuildTools
Microsoft.VisualStudio.Component.Workflow | Microsoft.VisualStudio.Component.PowerShell.Tools
| | Microsoft.VisualStudio.Component.Workflow.BuildTools
| | Microsoft.VisualStudio.Workload.OfficeBuildTools

**VS2019:**
Added | Not available
-- | --  
Microsoft.VisualStudio.Component.Sharepoint.Tools | Microsoft.VisualStudio.Component.TeamOffice.BuildTools
Microsoft.VisualStudio.Component.TeamOffice | Microsoft.VisualStudio.Component.Sharepoint.BuildTools
Microsoft.VisualStudio.Component.Workflow | Microsoft.VisualStudio.Component.Azure.MobileAppsSdk
| | Microsoft.VisualStudio.Component.Workflow.BuildTools
| | Microsoft.VisualStudio.Workload.OfficeBuildTools
| | Microsoft.Net.ComponentGroup.4.7.2.DeveloperTools
| | Component.UnityEngine.x86
| | Microsoft.Component.CookiecutterTools
| | Microsoft.VisualStudio.Component.VC.v141
#### Related issue:
https://github.com/actions/virtual-environments-internal/issues/802
## Check list
- [x] Related issue / work item is attached
- [ ] Tests are written (if applicable)
- [ ] Documentation is updated (if applicable)
- [ ] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/win/scripts/Installers/Windows2016/Install-VS2017.ps1 [hunk 1]
```diff
@@ -97,9 +28,7 @@ if ($instanceFolders -is [array])
     exit 1
 }
 
-$catalogContent = Get-Content -Path ($instanceFolders.FullName + '\catalog.json')
-$catalog = $catalogContent | ConvertFrom-Json
-$version = $catalog.info.id
+$version = Get-VisualStudioVersion
 $VSInstallRoot = "C:\Program Files (x86)\Microsoft Visual Studio\2017\$ReleaseInPath"
 Write-Host "Visual Studio version ${version} installed"
```
Comment (maxim-lobanov) [inline images/win/scripts/Installers/Windows2016/Install-VS2017.ps1:33]: Let's remove this comment and remove version getting at all

File: images/win/scripts/Installers/Windows2016/Install-VS2017.ps1 [hunk 2]
```diff
@@ -7,78 +7,9 @@ $ErrorActionPreference = "Stop"
 
 Import-Module -Name ImageHelpers -Force
```
Comment (maxim-lobanov) [inline images/win/scripts/Installers/Windows2016/Install-VS2017.ps1:9]:
```
$requiredComponents = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty workloads
$workloads = @("--allWorkloads", "--includeRecommended")
$workloads +=  $requiredComponents | ForEach-Object { "--add $_" }
$workloads += "--remove Component.CPython3.x64"
$workloadsArgument = [String]::Join(" ", $Workloads)
```

File: images/win/scripts/Tests/VisualStudio.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+        It "Devenv.exe" {
+            $expectedVersion = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty version
+            $devenvexePath = "${env:ProgramFiles(x86)}\Microsoft Visual Studio\${expectedVersion}\Enterprise\Common7\IDE\devenv.exe"
+            $devenvexePath | Should -Exist
+        }
+    }
+
+    Context "Visual Studio components" {
+        $visualStudioComponents = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty workloads
```
Comment (maxim-lobanov) [inline images/win/scripts/Tests/VisualStudio.Tests.ps1:16]: `$expectedComponents`

File: images/win/scripts/Tests/VisualStudio.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,26 @@
+Describe "Visual Studio" {
+    Context "Basic" {
+        It "Catalog.json" {
+            $catalogJsonPath = Get-VsCatalogJsonPath
+            $catalogJsonPath | Should -Exist
+        }
+
+        It "Devenv.exe" {
+            $expectedVersion = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty version
```
Comment (maxim-lobanov) [inline images/win/scripts/Tests/VisualStudio.Tests.ps1:9]: $expectedVisualStudioVersion

File: images/win/scripts/Installers/Windows2016/Install-VS2017.ps1 [hunk 5]
```diff
@@ -97,9 +28,7 @@ if ($instanceFolders -is [array])
     exit 1
 }
 
-$catalogContent = Get-Content -Path ($instanceFolders.FullName + '\catalog.json')
-$catalog = $catalogContent | ConvertFrom-Json
-$version = $catalog.info.id
+$version = Get-VisualStudioVersion
 $VSInstallRoot = "C:\Program Files (x86)\Microsoft Visual Studio\2017\$ReleaseInPath"
```
Comment (maxim-lobanov) [inline images/win/scripts/Installers/Windows2016/Install-VS2017.ps1:32]: Let's move this logic to helper function

File: images/win/scripts/Installers/Windows2019/Install-VS2019.ps1 [hunk 6]
```diff
@@ -6,104 +6,9 @@ $ErrorActionPreference = "Stop"
...
-              '--remove Component.CPython3.x64 ' + `
-              '--add Microsoft.VisualStudio.Workload.Universal ' + `
-              '--add Microsoft.VisualStudio.Workload.VisualStudioExtension ' + `
-              '--add Component.MDD.Linux ' + `
-              '--add Component.MDD.Linux.GCC.arm'
+$WorkLoads = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | `
+    Select-Object -ExpandProperty workloads | ForEach-Object { "-add $_" }
+$WorkLoads += "--remove Component.CPython3.x64"
 
 $ReleaseInPath = "Enterprise"
```
Comment (maxim-lobanov) [inline images/win/scripts/Installers/Windows2019/Install-VS2019.ps1:13]: $ReleaseInPath to toolset 
`Edition`

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 7]
```diff
@@ -465,3 +465,19 @@ function Extract-7Zip {
         exit 1
     }
 }
+
+function Get-VsCatalogJsonPath {
+    $vsProgramData = Get-Item -Path "C:\ProgramData\Microsoft\VisualStudio\Packages\_Instances"
+    $instanceFolders = Get-ChildItem -Path $vsProgramData.FullName
+    $catalogJsonPath = $instanceFolders.FullName + '\catalog.json'
```
Comment (maxim-lobanov) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:472]:
```
$instanceFolder = Get-Item "C:\ProgramData\Microsoft\VisualStudio\Packages\_Instances\*" | Select-Object -First 1
return Join-Path $instanceFolder.FullName "catalog.json"
```

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 8]
```diff
@@ -465,3 +465,19 @@ function Extract-7Zip {
...
+
+    return $catalogJsonPath
+}
+
+function Get-VisualStudioPath {
+    $visualStudioEdition = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty edition
+    $expectedVisualStudioVersion = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty version
```
Comment (maxim-lobanov) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:479]:
```
Param (
  [string]$Version,
  [string]$Edition
)
```

File: images/win/scripts/ImageHelpers/ImageHelpers.psm1 [hunk 9]
```diff
@@ -31,5 +31,7 @@ Export-ModuleMember -Function @(
     'Extract-7Zip'
     'Get-CommandResult'
     'Get-EnvironmentVariable'
-    'Invoke-PesterTests'
+    'Invoke-PesterTests',
```
Comment (maxim-lobanov) [inline images/win/scripts/ImageHelpers/ImageHelpers.psm1:34]: Let's remove `,` to be consistent

File: images/win/scripts/Installers/Windows2016/Install-VS2017.ps1 [hunk 10]
```diff
@@ -97,15 +28,14 @@ if ($instanceFolders -is [array])
     exit 1
 }
 
-$catalogContent = Get-Content -Path ($instanceFolders.FullName + '\catalog.json')
-$catalog = $catalogContent | ConvertFrom-Json
-$version = $catalog.info.id
-$VSInstallRoot = "C:\Program Files (x86)\Microsoft Visual Studio\2017\$ReleaseInPath"
-Write-Host "Visual Studio version ${version} installed"
+$expectedVisualStudioVersion = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty version
```
Comment (maxim-lobanov) [inline images/win/scripts/Installers/Windows2016/Install-VS2017.ps1:31]: `expectedVisualStudioVersion` such naming can confuse in installation file. Let's use just `visualStudioVersion`

File: images/win/scripts/Installers/Windows2016/Install-VS2017.ps1 [hunk 11]
```diff
@@ -5,86 +5,17 @@
 
 $ErrorActionPreference = "Stop"
 
-Import-Module -Name ImageHelpers -Force
+$requiredComponents = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty workloads
```
Comment (al-cheb) [inline images/win/scripts/Installers/Windows2016/Install-VS2017.ps1:8]: `(Get-ToolsetContent).visualStudio.workloads`

File: images/win/scripts/Installers/Windows2016/Install-VS2017.ps1 [hunk 12]
```diff
@@ -5,86 +5,17 @@
...
+$workLoads = @("--allWorkloads --includeRecommended")
```
Comment (al-cheb) [inline images/win/scripts/Installers/Windows2016/Install-VS2017.ps1:9]:
```
$requiredComponents = (Get-ToolsetContent).visualStudio.workloads | ForEach-Object { "--add $_" }
$workLoads = @(
	"--allWorkloads --includeRecommended"
	$requiredComponents
	"--remove Component.CPython3.x64"
)
```

File: images/win/scripts/Installers/Windows2016/Install-VS2017.ps1 [hunk 13]
```diff
@@ -5,86 +5,17 @@
...
-                '--add Microsoft.VisualStudio.Component.VC.Runtimes.x86.x64.Spectre '+ `
-                '--add Microsoft.VisualStudio.Component.VC.Runtimes.ARM.Spectre ' + `
-                '--add Microsoft.VisualStudio.Component.VC.Runtimes.ARM64.Spectre ' + `
-                '--add Microsoft.VisualStudio.Component.Workflow.BuildTools ' + `
-                '--add Microsoft.VisualStudio.Workload.Office ' + `
-                '--add Microsoft.VisualStudio.Workload.OfficeBuildTools '
-
-$ReleaseInPath = "Enterprise"
-$BootstrapperUrl = "https://aka.ms/vs/15/release/vs_${ReleaseInPath}.exe"
+$releaseInPath = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty edition
```
Comment (al-cheb) [inline images/win/scripts/Installers/Windows2016/Install-VS2017.ps1:14]: Add an additional variable for Get-ToolsetContent:
```
$toolset = Get-ToolsetContent
$releaseInPath = $toolset. visualStudio.edition
```

File: images/win/scripts/Installers/Windows2016/Install-VS2017.ps1 [hunk 14]
```diff
@@ -97,15 +28,14 @@ if ($instanceFolders -is [array])
     exit 1
 }
 
-$catalogContent = Get-Content -Path ($instanceFolders.FullName + '\catalog.json')
-$catalog = $catalogContent | ConvertFrom-Json
-$version = $catalog.info.id
-$VSInstallRoot = "C:\Program Files (x86)\Microsoft Visual Studio\2017\$ReleaseInPath"
-Write-Host "Visual Studio version ${version} installed"
+$visualStudioVersion = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty version
```
Comment (al-cheb) [inline images/win/scripts/Installers/Windows2016/Install-VS2017.ps1:31]: $toolset.visualStudio.version

File: images/win/scripts/Tests/VisualStudio.Tests.ps1 [hunk 15]
```diff
@@ -0,0 +1,28 @@
+Describe "Visual Studio" {
+    Context "Basic" {
+        It "Catalog.json" {
+            $catalogJsonPath = Get-VsCatalogJsonPath
+            $catalogJsonPath | Should -Exist
```
Comment (al-cheb) [inline images/win/scripts/Tests/VisualStudio.Tests.ps1:5]: Why don't we use a single line?

`Get-VsCatalogJsonPath | Should -Exist`

File: images/win/scripts/Installers/Windows2019/Install-VS2019.ps1 [hunk 16]
```diff
@@ -2,114 +2,20 @@
...
+$requiredComponents = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty workloads
+$workLoads = @("--allWorkloads --includeRecommended")
+$workLoads +=  $requiredComponents | ForEach-Object { "--add $_" }
+$workLoads += "--remove Component.CPython3.x64"
+$workLoadsArgument = [String]::Join(" ", $workLoads)
 
-$ReleaseInPath = "Enterprise"
-$BootstrapperUrl = "https://aka.ms/vs/16/release/vs_${ReleaseInPath}.exe"
+$releaseInPath = Get-ToolsetContent | Select-Object -ExpandProperty visualStudio | Select-Object -ExpandProperty edition
+$bootstrapperUrl = "https://aka.ms/vs/15/release/vs_${releaseInPath}.exe"
```
Comment (AlenaSviridenko) [inline images/win/scripts/Installers/Windows2019/Install-VS2019.ps1:15]: you are updating 16 -> 15 - is it correct?

Comment (dibir-magomedsaygitov) [reply images/win/scripts/Installers/Windows2019/Install-VS2019.ps1]: Thank you, changed back to 16.

**Comment by AlenaSviridenko** [comment]
Install-VS2017.ps1 and Install-VS2019.ps1 looks almost identical. Does it make sense to merge them and use `Test-IsWin19` condition in a couple of places?

**Comment by dibir-magomedsaygitov** [comment]
@AlenaSviridenko we plan to do it in scope of [this ](https://github.com/actions/virtual-environments-internal/issues/422)work-item.

**Pull request merged**